File: src/cxacru-info.c
/*
	cxacru-info - outputs cxacru status information from sysfs
	(also an example of obsessive checking of the return value
	and insane code to keep the layout of the output the same)

	Copyright ©2007 Simon Arlott

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License v2
	as published by the Free Software Foundation.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
	Or, point your browser to http://www.gnu.org/copyleft/gpl.html


	http://simon.arlott.org/sw/cxacru-info/
	$Id$
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MALLOC(ptr, size) do { \
		ptr = malloc(size); \
		if (ptr == NULL) { \
			perror("malloc"); \
			exit(3); \
		} \
	} while(0)

#define ZMALLOC(ptr, size) do { \
		MALLOC(ptr,size); \
		memset(ptr, 0, size); \
	} while(0)

#define OOPS(what) \
	fprintf(stderr, "%s at %s:%u\n", what, __FILE__, __LINE__)

#define PRINTF(format, args...) do { \
		int __ret = printf(format , ##args); \
		if (__ret <= 0) \
			exit(3); \
	} while(0)

#define FPRINTF(stream, format, args...) do { \
		int __ret = fprintf(stream, format , ##args); \
		if (__ret <= 0) \
			exit(3); \
	} while(0)

#define SNPRINTF(str, size, format, args...) do { \
		int __ret = snprintf(str, size, format , ##args); \
		if (__ret < 0) { \
			OOPS("snprintf failed"); \
			exit(3); \
		} else if (__ret >= size) { \
			OOPS("snprintf truncated"); \
			exit(3); \
		} \
	} while(0)

#define ERR_IF(expr, str) do { \
		if (expr) { \
			perror(str); \
			exit(1); \
		} \
	} while(0)

#define SYS_PATH "/sys/class/atm/"
#define ATM_DEVICES "/proc/net/atm/devices"
#define MAXLEN 255

unsigned long dev_num = -1;
unsigned long aal5[5];

enum { TXcount, TXerr, RXcount, RXerr, RXdrop };

char *cxacru(const char *file) {
	char filename[MAXLEN];
	char *data;

	ZMALLOC(data, MAXLEN + 1); // ensure string is always terminated
	SNPRINTF(filename, MAXLEN, SYS_PATH "cxacru%u/device/%s", dev_num, file);
	FILE *fd = fopen(filename, "r");
	ERR_IF(fd == NULL, filename);
	ERR_IF(fgets(data, MAXLEN, fd) == NULL, filename);
	ERR_IF(fclose(fd) != 0, filename);

	if (strlen(data) > 0) // remove newline
		data[strlen(data) - 1] = 0;
	return data;
}

char *strpad(char *str, int len) {
	if (strlen(str) >= len) {
		return str;
	} else {
		char *ret;

		MALLOC(ret, len + 1);
		memset(ret, ' ', len);
		memcpy(ret + (len - strlen(str)), str, strlen(str) + 1);
		free(str);

		return ret;
	}
}

char *intpad(unsigned int value, int len) {
	char *ret;

	MALLOC(ret, MAXLEN);
	SNPRINTF(ret, MAXLEN, "%u", value);

	return strpad(ret, len);
}

void find_atm_dev(long cxacru_num) {
	unsigned int num;
	int ret;
	FILE *fd = fopen(ATM_DEVICES, "r");

	ERR_IF(fd == NULL, ATM_DEVICES);
	ret = fscanf(fd, "%*[^\n][\n]");
	ERR_IF(ret < 0, ATM_DEVICES);

	while (6 == (ret = fscanf(fd,
			"%u cxacru %*s 0 ( %*ld %*ld %*ld %*ld %*ld ) 5 ( %ld %ld %ld %ld %ld ) [%*ld]%*[\n]",
			&num, &aal5[0], &aal5[1], &aal5[2], &aal5[3], &aal5[4]))) {
		if (cxacru_num < 0 || cxacru_num == num) {
			dev_num = num;
			break;
		}
	}
	ERR_IF(fclose(fd) != 0, ATM_DEVICES);
}

int main(int argc, char *argv[]) {
	char *modulation;

	if (argc > 2 || (argc == 2 && (!strcmp(argv[1], "-h") || !strcmp(argv[1], "--help")))) {
		PRINTF("Usage: %s [device num]\n", argv[0]);
		return 2;
	} else if (argc == 2 && (!strcmp(argv[1], "-v") || !strcmp(argv[1], "--version"))) {
		PRINTF("cxacru-info $Revision$ $Date$\n");
		return 0;
	}

	find_atm_dev(argc == 2 ? atol(argv[1]) : -1);
	if (dev_num == -1) {
		FPRINTF(stderr, "cxacru device not found\n");
		return 1;
	}

	PRINTF("                   Downstream     Upstream\n");
	PRINTF("\n");
	PRINTF("Line rate          %s"""" kbps    %s"""" kbps\n",
		strpad(cxacru("downstream_rate"), 6),
		strpad(cxacru("upstream_rate"), 6));
	PRINTF("Attenuation        %s"""" dB      %s"""" dB\n",
		strpad(cxacru("downstream_attenuation"), 6),
		strpad(cxacru("upstream_attenuation"), 6));
	PRINTF("Noise margin       %s"""" dB      %s"""" dB\n",
		strpad(cxacru("downstream_snr_margin"), 6),
		strpad(cxacru("upstream_snr_margin"), 6));
	PRINTF("Power                             %s"""" dBm/Hz\n",
		strpad(cxacru("transmitter_power"), 6));
	PRINTF("\n");
	PRINTF("AAL5 Frames    %s""""""""     %s""""""""\n",
		intpad(aal5[RXcount], 10), intpad(aal5[TXcount], 10));
	PRINTF("     Errors    %s""""""""     %s""""""""\n",
		intpad(aal5[RXerr], 10), intpad(aal5[TXerr], 10));
	PRINTF("     Dropped   %s""""""""\n",
		intpad(aal5[RXdrop], 10));
	PRINTF("\n");
	PRINTF("CRC errors     %s""""""""     %s""""""""\n",
		strpad(cxacru("downstream_crc_errors"), 10),
		strpad(cxacru("upstream_crc_errors"), 10));
	PRINTF("FEC errors     %s""""""""     %s""""""""\n",
		strpad(cxacru("downstream_fec_errors"), 10),
		strpad(cxacru("upstream_fec_errors"), 10));
	PRINTF("HEC errors     %s""""""""     %s""""""""\n",
		strpad(cxacru("downstream_hec_errors"), 10),
		strpad(cxacru("upstream_hec_errors"), 10));
	PRINTF("\n");
	PRINTF("Line status        %s\n", cxacru("line_status"));
	PRINTF("Link status        %s\n", cxacru("link_status"));
	modulation = cxacru("modulation");
	if (strlen(modulation) > 0)
	PRINTF("Modulation         %s\n", cxacru("modulation"));
	PRINTF("\n");
	PRINTF("MAC address        %s\n", cxacru("mac_address"));
	return 0;
}

File: src/cxacru-info.py
#!/usr/bin/env python
# encoding: utf-8
#
#	cxacru-info - outputs cxacru status information from sysfs
#
#	Copyright ©2009 Simon Arlott
#
#	This program is free software; you can redistribute it and/or
#	modify it under the terms of the GNU General Public License v2
#	as published by the Free Software Foundation.
#
#	This program is distributed in the hope that it will be useful,
#	but WITHOUT ANY WARRANTY; without even the implied warranty of
#	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#	GNU General Public License for more details.
#
#	You should have received a copy of the GNU General Public License
#	along with this program; if not, write to the Free Software
#	Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#	Or, point your browser to http://www.gnu.org/copyleft/gpl.html
#
#
#	http://simon.arlott.org/sw/cxacru-info/

from __future__ import print_function
import functools
import sys
import re

DEV_TYPE = "cxacru"
SYS_PATH = "/sys/class/atm/{type}{num}/device/{attr}"
ATM_DEVICES = "/proc/net/atm/devices"

ATM_LINE_PAT = re.compile('^ *(?P<itf>\d+) +(?P<type>\w+) +(?P<addr>\w)+(?P<aals>[^\t]+)\t\[(?P<refcnt>\d+)\]')
ATM_AALS_PAT = re.compile(' +(?P<num>\d+) +\( +(?P<tx_cnt>\d+) +(?P<tx_err>\d+) +(?P<rx_cnt>\d+) +(?P<rx_err>\d+) +(?P<rx_drop>\d+) +\)')

DS, US, XS = range(3)
DESC, TYPE, UNITS = range(3)
EXIT_SUCCESS, EXIT_FAILURE, EXIT_USAGE = range(3)

attributes = {
	# name:			desc,			type,	units
	"rate":			("Line rate",		XS,	" kbps"),
	"attenuation":		("Attenuation",		XS,	" dB"),
	"snr_margin":		("Noise margin",	XS,	" dB"),
	"transmitter_power":	("Power",		US,	" dBm/Hz"),
	"crc_errors":		("CRC errors",		XS,	""),
	"fec_errors":		("FEC errors",		XS,	""),
	"hec_errors":		("HEC errors",		XS,	""),
	"line_status":		("Line status",		DS,	None),
	"link_status":		("Link status",		DS,	None),
	"modulation":		("Modulation",		DS,	None),
	"mac_address":		("MAC address",		DS,	None)
}

def warn(msg):
	print("{0}: {1}".format(sys.argv[0], msg), file=sys.stderr)

def read_attr(name):
	try:
		with open(SYS_PATH.format(type=device["type"], num=device["itf"], attr=name), "r") as f:
			return f.read().rstrip("\n")
	except IOError as e:
		warn("Error reading {0}{1} stats".format(device["type"], device["itf"]))
		warn("IOError: {0}".format(e))
		exit(EXIT_FAILURE)

def find_dev(num):
	try:
		with open(ATM_DEVICES) as f:
			f.readline()
			for line in [ATM_LINE_PAT.match(line) for line in f]:
				if line is None:
					continue
				line = line.groupdict()

				if line["type"] == DEV_TYPE and (num == None or int(line["itf"]) == num):
					return line
	except IOError as e:
		warn("Error reading ATM devices")
		warn("IOError: {0}".format(e))
		exit(EXIT_FAILURE)
	return None

def get_aal(num):
	for aal in [aal.groupdict() for aal in ATM_AALS_PAT.finditer(device["aals"])]:
		if int(aal["num"]) == num:
			return aal
	return None

def print_line(desc, downstream, upstream):
	if downstream != "" and upstream == "":
		print("{0:15}{1}".format(desc, downstream))
	elif downstream != "":
		print("{0:15}{1:15}{2}".format(desc, downstream, upstream))

def format_attr(value, units):
	if units != None:
		return "{0:>10}{1}".format(value, units)
	else:
		return "{0:4}{1}".format("", value)

def print_attr(names):
	print()
	for name in names:
		attr = attributes[name]

		if attr[TYPE] == XS:
			print_line(attr[DESC],
				format_attr(read_attr("downstream_" + name), attr[UNITS]),
				format_attr(read_attr("upstream_" + name), attr[UNITS]))
		elif attr[TYPE] == DS:
			print_line(attr[DESC],
				format_attr(read_attr(name), attr[UNITS]), "")
		elif attr[TYPE] == US:
			print_line(attr[DESC],
				"", format_attr(read_attr(name), attr[UNITS]))

if "-h" in sys.argv[1:] or "--help" in sys.argv[1:] or len(sys.argv) > 2:
	print("Usage: {0} [device num]".format(sys.argv[0]))
	sys.exit(EXIT_USAGE)

if len(sys.argv) >= 2:
	if sys.argv[1] in ["-v", "--version"]:
		print("cxacru-info v0.9")
		sys.exit(EXIT_SUCCESS)
	else:
		try:
			num = int(sys.argv[1])
			if num < 0:
				raise ValueError
			device = find_dev(num)
		except ValueError:
			warn("ATM device number \"{0}\" is invalid".format(sys.argv[1]))
			sys.exit(EXIT_FAILURE)
		if device == None:
			warn("ATM device {0}{1} not found".format(DEV_TYPE, num))
			sys.exit(EXIT_FAILURE)
else:
	device = find_dev(None)
	if device == None:
		warn("no {0} ATM devices found".format(DEV_TYPE))
		sys.exit(EXIT_FAILURE)

aal5 = get_aal(5)
print_line("", format_attr("Downstream", None), format_attr("Upstream", None))
print_attr(["rate", "attenuation", "snr_margin", "transmitter_power"])
if aal5:
	print()
	print_line("AAL5 Frames", format_attr(aal5["rx_cnt"], ""), format_attr(aal5["tx_cnt"], ""))
	print_line("     Errors", format_attr(aal5["rx_err"], ""), format_attr(aal5["tx_err"], ""))
	print_line("     Dropped", format_attr(aal5["rx_drop"], ""), "")
print_attr(["crc_errors", "fec_errors", "hec_errors"])
print_attr(["line_status", "link_status", "modulation"])
print_attr(["mac_address"])
sys.exit(EXIT_SUCCESS)
